### ISSUE TITLE
multiple filter bug

### ISSUE THREAD

**Issue by agnellvj** [karmi/retire#2]
In the Filter#to_json method, it is possible to have an array of filters but when converting the filters to a hash, the last filter in the array will be the one that is used.

**Comment by karmi** [comment]
Good point, seems like an over-engineering on my part -- ES needs multiple filters wrapped in `bool` filter, which seems like a good way to accomplish what the API half-delivers. Any ideas?

**Comment by agnellvj** [comment]
I'm starting to wonder how the best way to implement this would be.  There are several different kinds filters.  This also applies to queries and facets.  Each has its own set of options and requirements.  Maybe a grammar with a custom DSL to build the queries, filters and facets.

```
filter do
  filter_group :and do
    filter_type :range do
      :field_name param1, param2
    end
    filter_type :prefix do
      :field_name param3
    end
  end
end
```

This should work with both "query" and "filtered" things.  For the bool filter, it would need another "do" layer for { must | should | must_not }.  Essentially we would have to go case by case and build the appropriate Search type up.

I'm not sure of an easier way to build the queries so that it implements the ES API.

**Comment by agnellvj** [comment]
Or another idea:

```
filter(:group => :and, :type => :range, "es_field_name" => {from: d1, to: d2})
filter(:bool, :must, :term =>  {:tag => "search_term"})
```

**Comment by karmi** [comment]
I like the block syntax you're proposing. I'll try to have a look into it.

**Comment by agnellvj** [comment]
I think it would work well with the infrastructure you have right now.  I think the block structure makes sense too given the ES API and how the Search API can be translated to JSON very easily.

I can try and help if you want.  I don't want to step on your toes or anything.

**Comment by agnellvj** [comment]
The more I think  about this, the more I think this would benefit from a more declarative style syntax.  It would make it easier to programmatically add filters and facets.  Kinda like building up a relation.  Something like:

```
query = Slingshot::Query.new(:query_string, "foo", opts_hash)
query.add_filter(:and, range: { from: START_DATE, to: END_DATE })
query.add_filter(:bool, must: { terms: { tags: ['bar', 'baz'] } })
```

The bool query is the only one that needs the "must | must_not | should" keyword so it's not worth making a special case for it.  This way you wouldn't have to worry about a dual interface as this should take care of all cases.

**Comment by karmi** [comment]
All good ideas. I recently implemented boolean queries in tire@d35e306876d5efbe9ed990b70ea57bc9a197b7a1. 

Multiple filters could be, in fact, defined with the `bool` filter, I've added a test in karmi/tire@01ba026.

I think I should expose the same API as for queries, eg. have a `boolean` method, with `must`, `should`, blocks... This way it'd be consistent at least.

I agree that the more declarative syntax, a la building an SQL query would be much better. But first we have to decouple the query building from the `perform` step...

**Comment by vburca** [comment]
Hello karmi, was this feature implemented eventually? I discovered this issue while looking for an answer to a related problem I have, that I posted on stackoverflow (http://stackoverflow.com/questions/17074279/filter-with-boolean-block-tire-elasticsearch)

Thanks!
